### PR TITLE
GH-34742: [Java] Split flight-sql-jdbc-driver to facilitate reuse

### LINKED ISSUE #363: [Java] Arrow-to-JDBC
ARROW-1780 reads a query from a JDBC data source and converts the ResultSet to an Arrow VectorSchemaRoot.  However, there is no built-in adapter for writing an Arrow VectorSchemaRoot back to the database.

ARROW-3966 adds JDBC field metadata:
- The Catalog Name
- The Table Name
- The Field Name
- The Field Type
  
  We can use this information to ask for the field information from the database via the [DatabaseMetaData](https://docs.oracle.com/javase/7/docs/api/java/sql/DatabaseMetaData.html) object.  We can then create INSERT or UPDATE statements based on the [list of primary keys](https://docs.oracle.com/javase/7/docs/api/java/sql/DatabaseMetaData.html#getPrimaryKeys(java.lang.String,%20java.lang.String,%20java.lang.String)) in the table:
- If the value in the VectorSchemaRoot corresponding to the primary key is NULL, insert that record into the database.
- If the value in the VectorSchemaRoot corresponding to the primary key is not NULL, update the existing record in the database.
  
  We can also perform the same data conversion in reverse based on the field types queried from the database.

**Reporter**: [Michael Pigott](https://issues.apache.org/jira/browse/ARROW-4144) / @mikepigott

<sub>**Note**: *This issue was originally created as [ARROW-4144](https://issues.apache.org/jira/browse/ARROW-4144). Please see the [migration documentation](https://github.com/apache/arrow/issues/14542) for further details.*</sub>

### LINKED ISSUE #34742: [Java] Split flight-sql-jdbc-driver to allow reuse of some internals
### Describe the enhancement requested

Continuing discussion from https://github.com/apache/arrow/issues/20730#issuecomment-1451989813 it may be useful to allow applications to use some of the JDBC driver internals to work with JDBC. They need to be split out as the driver JAR shades all classes. However, as these are not quite meant to be public APIs, it should be clear that the APIs are subject to change.

### Component(s)

Java

### PR DESCRIPTION
**Pull request by wgtmac**
### Rationale for this change

`flight-sql-jdbc-driver` module contains a set of utilities to implement JDBC support backed by arrow vectors. However, it employs maven shade plugin which is hard to make it a dependecy of external projects.

### What changes are included in this PR?

This patch splits `flight-sql-jdbc-driver` module into two separate modules: `flight-sql-jdbc-core` and `flight-sql-jdbc-driver`. `flight-sql-jdbc-core` contains almost every thing in the original module except the shade plugin. `flight-sql-jdbc-driver` only shades `flight-sql-jdbc-core`.

### Are these changes tested?

Make sure all tests pass.

### Are there any user-facing changes?

No.
* Closes: apache/arrow-java#363
* Closes: apache/arrow#34742
* GitHub Issue: apache/arrow#34742
* GitHub Issue: #34742

### CODE REVIEW

**Comment by wgtmac** [comment]
@lidavidm Could you please take a look? Thanks!

**Comment by lidavidm** [comment]
@davisusanibar or @lwhite1 would either of you be able to take a first pass?

Also, https://github.com/apache/arrow/blob/ce0d20c06aeb5c484888422cb4ce42e4d24c588c/dev/tasks/tasks.yml#L799-L803 needs to be updated with new artifacts

**Comment by wgtmac** [comment]
> https://github.com/apache/arrow/blob/ce0d20c06aeb5c484888422cb4ce42e4d24c588c/dev/tasks/tasks.yml#L799-L803
> 
> needs to be updated with new artifacts

Fixed `tasks.yml`, thanks @lidavidm!

BTW, though this PR looks very large by affecting hundreds of files, it contains only 200- lines of change as mostly are file relocation.

**Comment by assignUser** [comment]
@github-actions crossbow submit java-jars

**Comment by wgtmac** [comment]
@assignUser @kou  Could you please help rerun https://github.com/apache/arrow/actions/runs/4496207471/jobs/7910652606?pr=34678

The failure seems to be unrelated to this patch.

**Comment by wgtmac** [comment]
> Revision: [d382b46](https://github.com/apache/arrow/commit/d382b465be52e12d11aeb0cadcba008046432179)
> 
> Submitted crossbow builds: [ursacomputing/crossbow @ actions-5a76b30651](https://github.com/ursacomputing/crossbow/branches/all?query=actions-5a76b30651)
> 
> Task	Status
> java-jars	[Github Actions](https://github.com/ursacomputing/crossbow/actions/runs/4496253225/jobs/7910736314)

```
                  flight-sql-jdbc-core-12.0.0-SNAPSHOT-cyclonedx.json [     OK]
                   flight-sql-jdbc-core-12.0.0-SNAPSHOT-cyclonedx.xml [     OK]
                     flight-sql-jdbc-core-12.0.0-SNAPSHOT-javadoc.jar [     OK]
                     flight-sql-jdbc-core-12.0.0-SNAPSHOT-sources.jar [     OK]
                       flight-sql-jdbc-core-12.0.0-SNAPSHOT-tests.jar [     OK]
                             flight-sql-jdbc-core-12.0.0-SNAPSHOT.jar [     OK]
                             flight-sql-jdbc-core-12.0.0-SNAPSHOT.pom [     OK]
                flight-sql-jdbc-driver-12.0.0-SNAPSHOT-cyclonedx.json [     OK]
                 flight-sql-jdbc-driver-12.0.0-SNAPSHOT-cyclonedx.xml [     OK]
                   flight-sql-jdbc-driver-12.0.0-SNAPSHOT-javadoc.jar [PENDING]
                   flight-sql-jdbc-driver-12.0.0-SNAPSHOT-sources.jar [     OK]
                     flight-sql-jdbc-driver-12.0.0-SNAPSHOT-tests.jar [     OK]
                           flight-sql-jdbc-driver-12.0.0-SNAPSHOT.jar [     OK]
                           flight-sql-jdbc-driver-12.0.0-SNAPSHOT.pom [     OK]
Error: Process completed with exit code 1.
```

Let me fix `flight-sql-jdbc-driver-12.0.0-SNAPSHOT-javadoc.jar`

**Comment by wgtmac** [comment]
> Revision: [7067c95](https://github.com/apache/arrow/commit/7067c95c799e9c0d150cce938213d54f3759143a)
> 
> Submitted crossbow builds: [ursacomputing/crossbow @ actions-c5e181f11b](https://github.com/ursacomputing/crossbow/branches/all?query=actions-c5e181f11b)
> 
> Task	Status
> java-jars	[Github Actions](https://github.com/ursacomputing/crossbow/actions/runs/4497306667/jobs/7912785524)

```
Downloaded from apache: https://repo.maven.apache.org/maven2/com/google/guava/guava-parent/19.0/guava-parent-19.0.pom (9.9 kB at 1.2 MB/s)
[INFO] ------------------------------------------------------------------------
[INFO] Reactor Summary for Apache Arrow Java Root POM 12.0.0-SNAPSHOT:
[INFO] 
[INFO] Apache Arrow Java Root POM ......................... SUCCESS [ 45.069 s]
[INFO] Arrow Format ....................................... SUCCESS [ 16.880 s]
[INFO] Arrow Memory ....................................... SUCCESS [  5.681 s]
[INFO] Arrow Memory - Core ................................ SUCCESS [ 26.060 s]
[INFO] Arrow Memory - Unsafe .............................. SUCCESS [ 12.697 s]
[INFO] Arrow Memory - Netty ............................... SUCCESS [ 18.038 s]
[INFO] Arrow Vectors ...................................... FAILURE [10:15 min]
[INFO] Arrow Compression .................................. SKIPPED
[INFO] Arrow Tools ........................................ SKIPPED
[INFO] Arrow JDBC Adapter ................................. SKIPPED
[INFO] Arrow Plasma Client ................................ SKIPPED
[INFO] Arrow Flight ....................................... SKIPPED
[INFO] Arrow Flight Core .................................. SKIPPED
[INFO] Arrow Flight GRPC .................................. SKIPPED
[INFO] Arrow Flight SQL ................................... SKIPPED
[INFO] Arrow Flight SQL JDBC Driver Core .................. SKIPPED
[INFO] Arrow Flight SQL JDBC Driver ....................... SKIPPED
[INFO] Arrow Flight Integration Tests ..................... SKIPPED
[INFO] Arrow AVRO Adapter ................................. SKIPPED
[INFO] Arrow Algorithms ................................... SKIPPED
[INFO] Arrow Performance Benchmarks ....................... SKIPPED
[INFO] Arrow Java C Data Interface ........................ SKIPPED
[INFO] Arrow Orc Adapter .................................. SKIPPED
[INFO] Arrow Gandiva ...................................... SKIPPED
[INFO] Arrow Java Dataset ................................. SKIPPED
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  12:21 min
[INFO] Finished at: 2023-03-23T06:29:00Z
[INFO] ------------------------------------------------------------------------
Error:  Failed to execute goal org.apache.maven.plugins:maven-shade-plugin:3.1.1:shade (default) on project arrow-vector: Execution default of goal org.apache.maven.plugins:maven-shade-plugin:3.1.1:shade failed: Plugin org.apache.maven.plugins:maven-shade-plugin:3.1.1 or one of its dependencies could not be resolved: Failed to collect dependencies at org.apache.maven.plugins:maven-shade-plugin:jar:3.1.1 -> org.ow2.asm:asm:jar:6.0: Failed to read artifact descriptor for org.ow2.asm:asm:jar:6.0: Could not transfer artifact org.ow2.asm:asm:pom:6.0 from/to apache (https://repo.maven.apache.org/maven2/): Connection reset -> [Help 1]
Error:  
Error:  To see the full stack trace of the errors, re-run Maven with the -e switch.
Error:  Re-run Maven using the -X switch to enable full debug logging.
Error:  
Error:  For more information about the errors and possible solutions, please read the following articles:
Error:  [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/PluginResolutionException
Error:  
Error:  After correcting the problems, you can resume the build with the command
Error:    mvn <args> -rf :arrow-vector
Error: Process completed with exit code 1.
```

Failure is in `arrow-vector` which seems unrelated.

**Comment by kou** [comment]
I re-ran the job and the job passed. :-)

**Comment by wgtmac** [comment]
cc @sunchao :)

**Review by assignUser — COMMENTED** [review]
+1 on CI changes, can't comment on the rest :)

**Review by sunchao — APPROVED** [review]
Makes sense. It seems currently `flight-sql-jdbc-driver` just shade everything so it's difficult for downstream projects to manage dependencies (e.g., exclude certain dependency from `flight-sql-jdbc-driver`). 

I'm not a regular reviewer for Arrow-Java though so it's better for @lidavidm to make the final call.

File: java/flight/flight-sql-jdbc-core/pom.xml [hunk 1]
```diff
@@ -0,0 +1,184 @@
...
+        <artifactId>arrow-flight</artifactId>
+        <groupId>org.apache.arrow</groupId>
+        <version>12.0.0-SNAPSHOT</version>
+        <relativePath>../pom.xml</relativePath>
+    </parent>
+    <modelVersion>4.0.0</modelVersion>
+
+    <artifactId>flight-sql-jdbc-core</artifactId>
+    <name>Arrow Flight SQL JDBC Driver Core</name>
+    <description>(Contrib/Experimental) A JDBC driver based on Arrow Flight SQL.</description>
```
Comment (lidavidm) [inline java/flight/flight-sql-jdbc-core/pom.xml:26]: nit: can we update this to reflect that this isn't the actual driver?

**Review by lidavidm — APPROVED** [review]
LGTM. However, our testing of the driver artifact itself is nonexistent. We should consider setting up a job that takes the built driver JAR and runs it against the dremio-oss Docker container or similar.

File: java/flight/flight-sql-jdbc-core/pom.xml (hunk 1 above)
Comment (wgtmac) [reply java/flight/flight-sql-jdbc-core/pom.xml]: Slightly modified the description. Please correct me if there is any better description.

**Comment by wgtmac** [comment]
Do you have any idea why the `flight-sql-jdbc-core` test is easy to fail? @lidavidm 

https://github.com/apache/arrow/actions/runs/4527686135/jobs/7973768071?pr=34678

**Comment by lidavidm** [comment]
The test probably needs some debugging, but I have very little time for Java work these days (it's mostly on my personal time). There haven't been substantial contributions to the driver since its initial contribution.

**Comment by wgtmac** [comment]
> The test probably needs some debugging, but I have very little time for Java work these days (it's mostly on my personal time). There haven't been substantial contributions to the driver since its initial contribution.

I tried to debug it but cannot reproduce on my Mac. It seems to be unstable and now the tests are all green.

**Pull request merged**